### PR TITLE
8271202: C1: assert(false) failed: live_in set of first block must be empty

### PR DESCRIPTION
**Pull request by TheRealMDoerr**
I have written a checker which detects usage of the illegal phi function. In case of the reproducer provided in the JBS bug ("Reduced.java"), it finds the following and bails out:
```
invalidating local 8 because of type mismatch (new_value is NULL)
Bailing out because StoreIndexed (id 98) uses illegal phi (id 68)
```
I haven't checked why that node uses the illegal phi. That still seems to be a bug. Maybe there's a better solution to the underlying problem, but I hope my checker is useful to analyze bugs and to make C1 more resilient.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8271202](https://bugs.openjdk.java.net/browse/JDK-8271202): C1: assert(false) failed: live_in set of first block must be empty

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6683/head:pull/6683` \
`$ git checkout pull/6683`

Update a local copy of the PR: \
`$ git checkout pull/6683` \
`$ git pull https://git.openjdk.java.net/jdk pull/6683/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6683`

View PR using the GUI difftool: \
`$ git pr show -t 6683`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6683.diff">https://git.openjdk.java.net/jdk/pull/6683.diff</a>

</details>

### CODE REVIEW

**Comment by TheRealMDoerr** [comment]
Thanks for looking at it! I don't like my lengthy code so much, either. However, I haven't found a better place where we can check this. I don't think `LIRGenerator::move_to_phi` would be such a place because the direction is wrong. Defining node is the phi and the using one a StoreIndexed in this case. Wouldn't `LIRGenerator::move_to_phi` require it the other way around?
Note: Performance shouldn't be a concern as the search is only needed in very uncommon situations.

**Comment by neliasso** [comment]
How uncommon is it?

**Comment by TheRealMDoerr** [comment]
> How uncommon is it?

It usually never happens with standard configuration. In some cases when JVMTI capability `can_access_local_variables` is enabled (e.g. when using `-agentlib:jdwp=transport=dt_socket,address=localhost:8000,server=y,suspend=n`). C1 runs into it when compiling jvm98 `Compressor::compress` with JVMTI agent attached for example.
In addition, it happens when using -XX:+DeoptimizeALot which also extends live ranges, but that's only a debug option.

I don't think this PR is ready for integration, but it finds out interesting things which should not happen IMHO.

**Comment by veresov** [comment]
I took a deeper look and I added a comment to bug report.

The root cause seems to be because of the irreducible loops (and therefore an unusual block traversal order when inserting phis) the phi invalidation logic in try_merge() doesn't invalidate phis that have invalid locals as inputs. I've attached a drawing:
[8271202.pdf](https://github.com/openjdk/jdk/files/7739876/8271202.pdf). Notice that i54 = phi (i43, 96) is not invalidated even though 96 is illegal. Transitively, i43, it should be illegal too. I would propose that we add a check for that and bailout in move_phi(). 

Suggested fix: 

```
diff --git a/src/hotspot/share/c1/c1_LIRGenerator.cpp b/src/hotspot/share/c1/c1_LIRGenerator.cpp
index c064558b458..b386b541f89 100644
--- a/src/hotspot/share/c1/c1_LIRGenerator.cpp
+++ b/src/hotspot/share/c1/c1_LIRGenerator.cpp
@@ -963,6 +963,14 @@ void LIRGenerator::move_to_phi(PhiResolver* resolver, Value cur_val, Value sux_v
   Phi* phi = sux_val->as_Phi();
   // cur_val can be null without phi being null in conjunction with inlining
   if (phi != NULL && cur_val != NULL && cur_val != phi && !phi->is_illegal()) {
+    if (phi->is_local()) {
+      for (int i = 0; i < phi->operand_count(); i++) {
+        Value op = phi->operand_at(i);
+        if (op != NULL && op->type()->is_illegal()) {
+          bailout("illegal phi operand");
+        }
+      }
+    }
     Phi* cur_phi = cur_val->as_Phi();
     if (cur_phi != NULL && cur_phi->is_illegal()) {
       // Phi and local would need to get invalidated
```

**Comment by kelthuzadx** [comment]
I think we should at least fix/find such illegal use in state merging rather than LIR generation, that's far beyond where problem occurs

File: src/hotspot/share/c1/c1_Instruction.cpp [hunk 1]
```diff
@@ -835,7 +871,17 @@ bool BlockBegin::try_merge(ValueStack* new_state) {
           // In really rare cases we will bail out in LIRGenerator::move_to_phi.
           existing_phi->make_illegal();
           existing_state->invalidate_local(index);
-          TRACE_PHI(tty->print_cr("invalidating local %d because of type mismatch", index));
+          TRACE_PHI(tty->print_cr("invalidating local %d because of type mismatch%s",
+                                  index, new_value == NULL ? " (new_value is NULL)" : ""));
+          // Check if illegal phi gets used.
+          SearchUsageClosure search(existing_phi);
```
Comment (kelthuzadx) [inline src/hotspot/share/c1/c1_Instruction.cpp:877]: should we add something like "FIXME" comment to indicate this is a workaround and we need further investigation to solve it?

**Comment by veresov** [comment]
> I think we should at least fix/find such illegal use in state merging rather than LIR generation, that's far beyond where problem occurs

To find it you basically need another pass.  I don't want to introduce another pass just to find an extremely rare situation. Piggybacking on the LIR generation pass is the most cache-friendly place I can think of. In fact, there is already a related bailout in `move_to_phi()`.

I also don't think that this is necessarily even fixable. We artificially stretch the local liveness and that with conjunction with irreducible loops can create unmergable states. I think bailing out is appropriate. It is a very rare case.

**Comment by veresov** [comment]
Alright, we're tight on time with P3 for 18. I'm going to put out a corresponding PR.

**Comment by TheRealMDoerr** [comment]
Issue is fixed by https://github.com/openjdk/jdk18/commit/54b800d56d6bc86676722ad96e87b8344606bcb7. I wonder if we still have a potential problem when an illegal phi gets used. That can still happen (not in the provided `test()` method). E.g. `If (id 73) uses illegal phi (id 68)` in jvm98 `Compressor::compress`.

**Pull request closed without merging**